doc/warm-reboot/NSF_Manager_HLD.md

```diff
@@ -0,0 +1,280 @@
+
+
+##### Phase 2: State Verification (Optional)
+
+
+
+![alt_text](img/state-verification.png)
+
+Since the switch is in quiescent state, this will be the final state of the switch before reboot. NSF Manager will trigger state verification to ensure that the switch is in a consistent state. Reconciling from an inconsistent state might cause traffic loss and thus it is important to ensure that the switch is in a consistent state before warm rebooting. NSF Manager will abort warm reboot operation if state verification fails.
+
+
+##### Phase 3: Trigger Checkpointing
+
+
+
+![alt_text](img/checkpoint.png)
+
+NSF Manager will send checkpoint notification to all registered components and wait for only to those components that set _checkpoint = true_ to trigger checkpointing i.e. save internal states to either the DB or persistent file system. Checkpointing after state verification is successful ensures that the switch is reconciling from a consistent state. NSF Manager will wait for a period of time for the components to update STATE DB with their checkpointing status after which it will abort the warm reboot operation.
+
+
+##### Phase 4: Prepare and Perform Reboot
+
+
+
+![alt_text](img/phase-4-shutdown.png)
+
+At this point, the DB will be backed up, containers will be shutdown, the platform will be prepared for reboot and the switch will be rebooted.
+
+
+##### Application Shutdown Optimization
+
+
+
+![alt_text](img/shutdown-optimization.png)
+
+Higher layer applications are generally independent and their shutdown is not dependent on quiescence of other switch stack components. For example, upon receiving a freeze notification Teamd can stop exchanging LACP packets, perform state verification (optional) and checkpoint LACP states i.e. transition from Phase 1 to Phase 3 without waiting for NSF Manager to send further notifications. The design allows applications to transition through these phases independently as long as they continue to update their warmboot state in STATE DB. NSF Manager will monitor these states and will handle the applications’ phase transitions. In such a scenario, applications need to set _freeze = true_ and _checkpoint = false_ and thus the freeze notification will result in the application to transition from Phase 1 to Phase 3. 
```

Comment:
> The design allows applications to transition through these phases independently

But in warm shutdown we do need dependent transition. 
`independent` transition can cause confusion and open gates for multiple race conditions.

Reply:
Phase 1 ensures that the switch stack components continue to serve requests from other components. Also, having a state machine during shutdown helps alleviate the race conditions. I would be happy to look at individual examples of race conditions and check if this design is able to handle them.